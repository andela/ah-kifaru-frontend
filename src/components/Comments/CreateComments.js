import React from 'react';
import PropTypes from 'prop-types';

export const CreateComments = ({
  handleCommentSubmit,
  handleInputFieldChange,
  content,
  isAuthenticated,
  contentError,
  handleClear
}) => (
  <section>
    <div className="flex flex-col w-full m-auto  mb-10 mt-10">
      <h2 className="mb-3 font-bold text-2xl uppercase">comments</h2>
      <form
        onSubmit={handleCommentSubmit}
        data-testid="submit-form"
        arai-label="comment"
      >
        <div className="flex flex-col md:flex-row w-full ">
          <div className="w-full md:pr-6 pr-0">
            <textarea
              data-testid="content"
              className="w-full h-full p-3 border-solid border-gray-200 border-2 "
              placeholder="Enter Comment..."
              name="content"
              value={content}
              onChange={handleInputFieldChange}
              disabled={!isAuthenticated}
            />
            {!isAuthenticated ? (
              <p className="text-center text-red-600 font-medium text-xl">
                You need to login to post a comment!!!
              </p>
            ) : (
              ''
            )}
            {!!contentError && (
              <span className="text-red-600 text-base">{contentError}</span>
            )}
          </div>
          <div
            data-testid="commentButtons"
            className="md:w-1/5 w-full flex flex-row justify-between md:flex-col md:justify-start"
          >
            <button
              data-testid="button"
              type="button"
              className="bg-gray-500 hover:bg-gray-600 text-black p-2 md:p-4 rounded mb-5 md:w-full w-1/2 md:ml-0 mr-5"
              onClick={handleClear}
              disabled={!isAuthenticated}
            >
              CLEAR
            </button>
            <button
              data-testid="submit"
              type="submit"
              className="bg-red-500 hover:bg-red-600 text-white p-2 md:p-4 rounded mb-5 md:w-full w-1/2 md:ml-0 ml-5"
              disabled={!isAuthenticated}
            >
              SEND
            </button>
          </div>
        </div>
      </form>
    </div>
  </section>
);

CreateComments.propTypes = {
  content: PropTypes.string,
  contentError: PropTypes.string,
  handleClear: PropTypes.func,
  handleCommentSubmit: PropTypes.func,
  handleInputFieldChange: PropTypes.func,
  isAuthenticated: PropTypes.bool.isRequired
};

CreateComments.defaultProps = {
  handleInputFieldChange: PropTypes.func.isRequired,
  handleCommentSubmit: PropTypes.func.isRequired,
  content: PropTypes.isRequired,
  contentError: PropTypes.isRequired,
  handleClear: PropTypes.func.isRequired
};

export default CreateComments;
