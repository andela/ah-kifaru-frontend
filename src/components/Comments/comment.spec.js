import React from 'react';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import CommentsCard from '@components/Comments/index';
import { BrowserRouter, Switch } from 'react-router-dom';
import { Provider } from 'react-redux';
import {
  render,
  fireEvent,
  waitForDomChange,
  cleanup
} from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { CreateComments } from './CreateComments';

const mockStore = configureMockStore([thunk]);
let store;
const initialSate = {
  comments: [],
  error: null,
  status: 'rest'
};

const props = {
  isAuthenticated: true,
  match: {
    params: {
      articleId: 55
    }
  },
  getComment: jest.fn()
};

const renderWithEnzymes = state => {
  store = mockStore({
    authReducer: {
      error: null,
      isAuthenticated: false,
      user: {},
      status: 'rest'
    },
    commentReducer: {
      comments: [],
      error: null,
      status: 'rest'
    },
    commentsCard: {
      ...state
    }
  });
  return mount(
    <Provider store={store}>
      <BrowserRouter>
        <Switch>
          <CommentsCard {...props} />
        </Switch>
      </BrowserRouter>
    </Provider>
  );
};

const renderWithRTL = state => {
  store = mockStore({
    authReducer: {
      error: null,
      isAuthenticated: false,
      user: {},
      status: 'rest'
    },
    commentReducer: {
      comments: [],
      error: null,
      status: 'rest'
    },
    commentsCard: {
      ...state
    }
  });
  return {
    ...render(
      <Provider store={store}>
        <BrowserRouter>
          <Switch>
            <CommentsCard {...props} />
          </Switch>
        </BrowserRouter>
      </Provider>
    ),
    store
  };
};

describe('CommentsCard', () => {
  afterEach(() => {
    jest.clearAllMocks();
    store = null;
  }, cleanup);

  it('should render comment buttons', () => {
    const renderComponent = () => {
      return render(
        <BrowserRouter>
          <Switch>
            <CreateComments {...props} />
          </Switch>
        </BrowserRouter>
      );
    };
    const { getByTestId } = renderComponent(initialSate);
    expect(getByTestId('button')).toBeDefined();
    expect(getByTestId('submit')).toBeDefined();
    expect(getByTestId('commentButtons')).toBeDefined();
  });

  it('should have a textfield', () => {
    const wrapper = renderWithEnzymes();
    const field = wrapper.find('textarea').first();
    expect(field.exists()).toBe(true);
    expect(field.props().name).toEqual('content');
  });

  it('should have a send submit button', () => {
    const wrapper = renderWithEnzymes();
    const button = wrapper.find("button[type='submit']");
    expect(button.exists()).toBe(true);
    expect(button.text()).toEqual('SEND');
  });

  it('should have a clear button', () => {
    const wrapper = renderWithEnzymes();
    const button = wrapper.find("button[type='button']");
    expect(button.exists()).toBe(true);
    expect(button.text()).toEqual('CLEAR');
  });

  it('should render CommentsCard Component', () => {
    const wrapper = renderWithEnzymes();
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should update comment field on change', () => {
    const wrapper = renderWithEnzymes();
    const commentInput = wrapper.find("textarea[name='content']");
    commentInput.simulate('change', {
      persist: () => {},
      target: {
        name: 'content',
        value: 'I love you!'
      }
    });
    expect(commentInput.html()).toMatch('I love you!');
  });
  it('should post comment field on submit', () => {
    const wrapper = renderWithEnzymes();
    const event = {
      preventDefault: jest.fn()
    };
    const commentInput = wrapper.find("textarea[name='content']");
    commentInput.simulate('click', {
      persist: () => {},
      target: {
        name: 'content',
        value: 'I love you!'
      }
    });
    wrapper.find('form').simulate('submit', event);
    expect(wrapper.find('CommentsCard').state('content')).toEqual('');
  });

  it('should clear the textarea on clear button', () => {
    const wrapper = renderWithEnzymes();
    const commentInput = wrapper.find("textarea[name='content']");
    commentInput.simulate('change', {
      persist: () => {},
      target: {
        name: 'content',
        value: ''
      }
    });
    wrapper.find('form').simulate('submit');
    expect(commentInput.html()).toMatch('');
    expect(wrapper.find('CommentsCard').state('content')).toEqual('');
  });

  it('should submit a valid form', async () => {
    const { getByTestId, findByText } = renderWithRTL(initialSate);

    const commentInput = getByTestId('content');
    fireEvent.change(commentInput, {
      target: { value: 'I love you' }
    });

    const formNode = getByTestId('submit-form');
    fireEvent.submit(formNode);

    await waitForDomChange(() => {
      const send = findByText('SEND');
      expect(send).toBeTruthy();
      expect(props.handleCommentSubmit).toHaveBeenCalled();
    });
  });
});
