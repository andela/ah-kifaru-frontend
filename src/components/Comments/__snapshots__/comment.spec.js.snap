// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommentsCard should render CommentsCard Component 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <BrowserRouter>
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <Switch>
        <Connect(CommentsCard)
          computedMatch={
            Object {
              "isExact": true,
              "params": Object {},
              "path": "/",
              "url": "/",
            }
          }
          getComment={[MockFunction]}
          isAuthenticated={true}
          location={
            Object {
              "hash": "",
              "pathname": "/",
              "search": "",
              "state": undefined,
            }
          }
          match={
            Object {
              "params": Object {
                "articleId": 55,
              },
            }
          }
        >
          <CommentsCard
            comments={Array []}
            computedMatch={
              Object {
                "isExact": true,
                "params": Object {},
                "path": "/",
                "url": "/",
              }
            }
            getComment={[Function]}
            isAuthenticated={false}
            location={
              Object {
                "hash": "",
                "pathname": "/",
                "search": "",
                "state": undefined,
              }
            }
            match={
              Object {
                "params": Object {
                  "articleId": 55,
                },
              }
            }
            postComment={[Function]}
            user={Object {}}
          >
            <div
              className="w-4/6 mx-auto"
            >
              <CreateComments
                comments={Array []}
                content=""
                contentError=""
                handleClear={[Function]}
                handleCommentSubmit={[Function]}
                handleInputFieldChange={[Function]}
                isAuthenticated={false}
                status="rest"
              >
                <section>
                  <div
                    className="flex flex-col w-full m-auto  mb-10 mt-10"
                  >
                    <h2
                      className="mb-3 font-bold text-2xl uppercase"
                    >
                      comments
                    </h2>
                    <form
                      arai-label="comment"
                      data-testid="submit-form"
                      onSubmit={[Function]}
                    >
                      <div
                        className="flex flex-col md:flex-row w-full "
                      >
                        <div
                          className="w-full md:pr-6 pr-0"
                        >
                          <textarea
                            className="w-full h-full p-3 border-solid border-gray-200 border-2 "
                            data-testid="content"
                            disabled={true}
                            name="content"
                            onChange={[Function]}
                            placeholder="Enter Comment..."
                            value=""
                          />
                          <p
                            className="text-center text-red-600 font-medium text-xl"
                          >
                            You need to login to post a comment!!!
                          </p>
                        </div>
                        <div
                          className="md:w-1/5 w-full flex flex-row justify-between md:flex-col md:justify-start"
                          data-testid="commentButtons"
                        >
                          <button
                            className="bg-gray-500 hover:bg-gray-600 text-black p-2 md:p-4 rounded mb-5 md:w-full w-1/2 md:ml-0 mr-5"
                            data-testid="button"
                            disabled={true}
                            onClick={[Function]}
                            type="button"
                          >
                            CLEAR
                          </button>
                          <button
                            className="bg-red-500 hover:bg-red-600 text-white p-2 md:p-4 rounded mb-5 md:w-full w-1/2 md:ml-0 ml-5"
                            data-testid="submit"
                            disabled={true}
                            type="submit"
                          >
                            SEND
                          </button>
                        </div>
                      </div>
                    </form>
                  </div>
                </section>
              </CreateComments>
            </div>
          </CommentsCard>
        </Connect(CommentsCard)>
      </Switch>
    </Router>
  </BrowserRouter>
</Provider>
`;
